RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,266 @@
+# Unit checking
+
+Unit checking is the process of inferring the units of Modelica variables, and ensuring the variables are used consistently in expressions and equations.
+
+It takes place after any evaluation of evaluable parameters and no evaluable parameters shall be evaluated during this process.
+(This must be true as long as tools allow the users to opt-out from unit-checking.)
+
+It is performed by using a type inference algorithm, albeit adapted to work with units.
+The algorithm proposed here is based on the Hindley-Milner algorithm.
+At a high level the algorithm collects a set of constraints on the units of expressions.
+It extracts a sequence of substitutions from these constraints which once applied constructs the unit of a given expression.
```

Comment:
Here's an example with mutually dependent unit variables:
```
model MutuallyDependenUnits
  parameter Real k1(unit = "m") = 1;
  parameter Real k2(unit = "s") = 1;
  Real x;
  Real y;
equation
  x^2 * y^3 = k1;
  x^3 * y^2 = k2;
end MutuallyDependenUnits;
```
This is solved by:
```
x.unit = "s(3/5).m-(2/5)"
y.unit = "m(3/5).s-(2/5)"
```

After your explanation of what the kernel implementation should be doing, I would have expected it to find the solution, but for some reason it doesn't.

Anyway, I think I now see in what way the "application of the sequence to any unit expression" could be understood as just a less efficient way of the idea I'm trying to convey here.  I mean, isn't a dictionary of solutions (possibly _unknown_) for all unit variables the most efficient structure for determining the unit of an arbitrary expression, and at the same time the structure of primary interest to us for getting the inferred units of Modelica variables?  With this in mind, I would only apply "the sequence" to the substitutions themselves (in addition to the constraints remaining to be processed).

When a constraint is processed and a substitution obtained, the substitution will have no impact on later obtained substitutions, since the eliminated variable will have been removed from the corresponding constraints.  Hence, if a new substitution is applied immediately to previously obtained substitutions as well as remaining constraints, the total number of places where a substitution is applied cannot increase but only decrease each time a satisfied constraint without unit variables is removed.  Thus, there also seems to be a performance gain in immediate application of constraints instead of keeping track of a sequence.

Applied to `MutuallyDependenUnits`, one would start with the constraints:
```
k1.unit ~ "m"
k2.unti ~ "s"
x.unit^2 * y.unit^3 ~ k1.unit
x.unit^3 * y.unit^2 ~ k2.unit
```
After processing the first two constraints we're in the state:
```
k1.unit -> "m"
k2.unti -> "s"
x.unit^2 * y.unit^3 ~ "m"
x.unit^3 * y.unit^2 ~ "s"
```
Here, the approach in the notebook would solve simultaneously for `x.unit` and `y.unit`, but let's do it sequentially instead.  Solve `x.unit^2 * y.unit^3 ~ "m"` with respect to `x.unit` first:
```
k1.unit -> "m"
k2.unti -> "s"
x.unit -> "m"^(1/2) * y.unit^-(3/2)
"m(3/2)" * y.unit^-(5/2) ~ "s"
```
Finally, solving for `y.unit`, we get (after applying the substitution to the previously obtained constraint):
```
k1.unit -> "m"
k2.unti -> "s"
x.unit -> "s(3/5).m-(2/5)"
y.unit -> "s-(2/5).m(3/5)"
```

Do you agree that the algorithm should be capable (and allowed) to conclude this?

Reply:
Yes, I fully agree, this is how it is implemented and I consider it a bug that it isn't working.

Reply:
The bug should be fixed in our implementation.